### PR TITLE
Add Adot adapter

### PR DESCRIPTION
**Pull request by Giudici-a**
Hi,

Following the #1599, i moved the converted version of the adapter here to make it cleaner.

### CODE REVIEW

File: adapters/adot/adot_test.go [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+	"github.com/prebid/prebid-server/adapters/adapterstest"
+	"github.com/prebid/prebid-server/config"
+	"github.com/prebid/prebid-server/openrtb_ext"
+	"github.com/stretchr/testify/assert"
+	"io/ioutil"
+	"strings"
+	"testing"
+)
+
+var jsonBidReq = getJsonByteForTesting("./static/adapter/adot/parallax_request_test.json")
```
Comment (SyntaxNode) [inline adapters/adot/adot_test.go:16]: This path is already scoped to this adapter package. Why the need for so many directory levels? It's also likely not necessary as you should be able to use our standard json test framework, no?

File: adapters/adot/adot_test.go [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+//Test the media type for a bid response
+func TestMediaTypeForBid(t *testing.T) {
+	_, err := getMediaTypeForBid(nil, nil)
+
+	assert.Error(t, err)
+
+	var reqBanner, reqVideo *openrtb.BidRequest
+	var bidBanner, bidVideo *openrtb.Bid
+
+	err1 := json.Unmarshal(getJsonByteForTesting("./static/adapter/adot/parallax_request_test.json"), &reqBanner)
```
Comment (SyntaxNode) [inline adapters/adot/adot_test.go:107]: Can you use our existing json test framework here?

File: adapters/adot/adottest/supplemental/parallax.json [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+            }
+          ],
+          "bidid": "5778926625248726496",
+          "cur": "USD"
+        }
+      }
+    }
+  ],
+
+  "expectedBids": [
```
Comment (SyntaxNode) [inline adapters/adot/adottest/supplemental/parallax.json:80]: This is an incorrect name and is ignored by our test framework. Should be `expectedBidResponses`. See: https://docs.prebid.org/prebid-server/developers/add-new-bidder-go.html#adapter-code-tests

File: adapters/adot/adot.go [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+		Uri:     a.endpoint,
+		Body:    reqJSON,
+		Headers: headers,
+	}}, errors
+}
+
+// MakeBids unpacks the server's response into Bids.
+func (a *AdotAdapter) MakeBids(internalRequest *openrtb.BidRequest, externalRequest *adapters.RequestData, response *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	if response.StatusCode == http.StatusNoContent {
+		return nil, nil
```
Comment (SyntaxNode) [inline adapters/adot/adot.go:60]: Please add test coverage these error code conditions.

**Comment by Giudici-a** [comment]
Thanks you for your responses, i made the changes about the returns that you asked.

File: adapters/adot/adot.go [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+	if err := json.Unmarshal(bid.Ext, &impExt); err == nil {
+		switch impExt.Adot.MediaType {
+		case string(openrtb_ext.BidTypeBanner):
+			return openrtb_ext.BidTypeBanner, nil
+		case string(openrtb_ext.BidTypeVideo):
+			return openrtb_ext.BidTypeVideo, nil
+		case string(openrtb_ext.BidTypeAudio):
+			return openrtb_ext.BidTypeAudio, nil
+		case string(openrtb_ext.BidTypeNative):
+			return openrtb_ext.BidTypeNative, nil
```
Comment (bsardo) [inline adapters/adot/adot.go:110]: Could you add a native JSON test please?

File: adapters/adot/adot.go [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+
+	if response.StatusCode != http.StatusOK {
+		return nil, []error{&errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Unexpected status code: %d. Run with request.debug = 1 for more info", response.StatusCode),
+		}}
+	}
+
+	var bidResp openrtb.BidResponse
+	if err := json.Unmarshal(response.Body, &bidResp); err != nil {
+		return nil, []error{err}
```
Comment (bsardo) [inline adapters/adot/adot.go:75]: We're missing coverage here. See `adapters/connectad/connectadtest/supplemental/badresponse.json` as an example.

File: adapters/adot/adot_test.go [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+	assert.Error(t, err)
+
+	byteInvalid, _ := json.Marshal(&adotBidExt{Adot: bidExt{"invalid"}})
+	_, err = getMediaTypeForBid(&openrtb.Bid{Ext: json.RawMessage(byteInvalid)})
+
+	assert.Error(t, err)
+}
+
+//Test the bid response when the bidder return a status code 204
+func TestBidResponseNoContent(t *testing.T) {
```
Comment (bsardo) [inline adapters/adot/adot_test.go:38]: You could also achieve this coverage using the JSON test framework.

**Review by bsardo — CHANGES_REQUESTED** [review]

**Comment by Giudici-a** [comment]
Hi, thanks again for your responses.
I pushed in the previous commit all the corrective request that you told me.

File: adapters/adot/adot.go [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+
+// MakeBids unpacks the server's response into Bids.
+// The bidder return a status code 204 when it cannot delivery an ad.
+func (a *adapter) MakeBids(internalRequest *openrtb.BidRequest, externalRequest *adapters.RequestData, response *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	if response.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+
+	if response.StatusCode == http.StatusBadRequest {
+		return nil, []error{&errortypes.BadInput{
```
[the remark this answers is not in the document]
Comment (SyntaxNode) [reply adapters/adot/adot.go:62]: This specific error condition still looks to be untested.

**Review by SyntaxNode — COMMENTED** [review]
Thank you for taking our feedback into account and updating the PR. Looks good to me, just except test coverage of the 400 error code from your bidding server.

**Comment by Giudici-a** [comment]
I did the last think to test the status code 400, hope everything is good now.
I pushed it again in the previous commit, cause i think it's cleaner like that.
Thanks for all your feedback.

File: adapters/adot/adottest/supplemental/simple-audio.json [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+              }
+            }
+          ]
+        }
+      },
+      "mockResponse": {
+        "status": 200,
+        "body": {
+          "status": 200,
+          "body": {
```
Comment (bsardo) [inline adapters/adot/adottest/supplemental/simple-audio.json:43]: It looks like you've got `status` and `body` twice. Can you fix so there's only one occurrence of each?

File: adapters/adot/adot.go (hunk 6 above)
Comment (bsardo) [reply adapters/adot/adot.go:75]: We're still missing coverage here for this error condition.

**Review by bsardo — CHANGES_REQUESTED** [review]
This is looking good, just two more comments. If you could make these changes and push them up in a new commit that'd be great. It would help me review the delta faster.

**Review by SyntaxNode — APPROVED** [review]

**Comment by SyntaxNode** [comment]
Looks great. Thank you for addressing our feedback. Can you please post a link to your [documentation PR](https://docs.prebid.org/prebid-server/developers/add-new-bidder-go.html#user-documentation)? Then we should be all set.

**Review by bsardo — APPROVED** [review]
This looks good. @Giudici-a we're just waiting for you to open a doc PR before merging.

**Comment by SyntaxNode** [comment]
@Giudici-a we approved this adapter, but you need to open a PR on the https://github.com/prebid/prebid.github.io project to change https://github.com/prebid/prebid.github.io/blob/master/dev-docs/bidders/adot.md per the [docs](https://docs.prebid.org/prebid-server/developers/add-new-bidder-go.html#user-documentation).

Specifically, you need to add `pbs: true` and a new table showing that placementId and parallax are the parameters for the PBS adapter - since the PBJS adapter has a different set of parameters. We appreciate the consistency with placementId though.

**Comment by bretg** [comment]
docs PR https://github.com/prebid/prebid.github.io/pull/2609

**Pull request merged**